File: 2023/src/Day1/Part2.java
package Day1;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Part2 {

    enum Numbers {
        zero,
        one,
        two,
        three,
        four,
        five,
        six,
        seven,
        eight,
        nine
    }

    static String numbers = "0123456789";

    static int indexOf(String line, String num) {
        int index = line.indexOf(num);
        if (index < 0) {
            return Integer.MAX_VALUE;
        }

        return index;
    }

    static String convertToInt(String num) {
        try {
            Integer.parseInt(num);
            return num;
        } catch (NumberFormatException e) {
            return "" + Numbers.valueOf(num).ordinal();
        }
    }

    public static void main(String[] args) throws IOException {
        String file = "C:\\Users\\User\\Downloads\\advent.txt";

        int total = 0;
        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
            for(String line; (line = reader.readLine()) != null; ) {
                List<String> numStrings = Stream.concat(
                        Arrays.stream(Numbers.values()).map(Enum::toString),
                        Arrays.stream(numbers.split(""))
                ).collect(Collectors.toList());

                final String temp = line;
                String firstDigit = numStrings.stream()
                        .reduce((num1, num2) -> indexOf(temp, num1) < indexOf(temp, num2) ? num1 : num2)
                        .orElseThrow(RuntimeException::new);

                String secondDigit = numStrings.stream()
                        .reduce((num1, num2) -> temp.lastIndexOf(num1) > temp.lastIndexOf(num2) ? num1 : num2)
                        .orElseThrow(RuntimeException::new);

                total += Integer.parseInt(convertToInt(firstDigit) + convertToInt(secondDigit));
            }
        }

        System.out.println(total);
    }
}

File: src/2024/Day09/Part2.py
memory = []
with open("/home/user/Downloads/advent.txt") as file:
    line = file.readline().strip()

    for i in range(len(line)):
        num = int(line[i])
        memory.append((i // 2, num) if i % 2 == 0 else (".", num))

reverse = memory[::-1]

for i in range(0, len(reverse), 2):
    k = memory.index(reverse[i])
    for j in range(1, k, 2):
        if reverse[i][1] <= memory[j][1]:
            k -= 1
            memory[k] = (memory[k][0], sum(x[1] for x in memory[k:k + 3]))
            memory[k + 1:k + 3] = []

            memory[j] = (memory[j][0], memory[j][1] - reverse[i][1])
            memory[j:j + 1] = [(".", 0), reverse[i], memory[j]]
            break

memory = [val for val, num in memory for _ in range(num)]
print(sum(memory[i] * i for i in range(len(memory)) if memory[i] != "."))
